src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/FileBasedProgramsProjectSystem.cs

```diff
@@ -77,12 +77,22 @@ public async ValueTask<bool> IsMiscellaneousFilesDocumentAsync(TextDocument docu
         //   1.  The document is a primordial document (either not loaded yet or doesn't support design time build) - it will be in the misc files workspace.
         //   2.  The document is loaded as a canonical misc file - these are always in the misc files workspace.
         //   3.  The document is loaded as a file based program - then it will be in the main workspace where the project path matches the source file path.
+
+        // NB: The FileBasedProgramsProjectSystem uses the document file path (the on-disk path) as the projectPath in 'IsProjectLoadedAsync'.
+        var isLoadedAsFileBasedProgram = document.FilePath is { } filePath && await IsProjectLoadedAsync(filePath, cancellationToken);
+
+        // If this document has a file-based program syntactic marker, but we aren't loading it in a file-based programs project,
+        // we need the caller to remove and re-add this document, so that it gets put in a file-based programs project instead.
```

Comment:
update this comment to explain how this works (as fddiscussed offline).  will tackle improving the clarity of this code in one of the followups I am working on